### PR TITLE
feat(python): add `@py_class` decorator for Python-defined FFI dataclasses

### PR DESCRIPTION
**Pull request by junrushao**
## Summary

Add `@py_class`, a new decorator that lets users define TVM FFI Object types entirely in Python with dataclass-style field annotations. This complements `@c_class` (which wraps C++-backed types) by enabling pure-Python FFI types that participate in the same object system -- type registry, reference counting, cross-language serialization, and packed-function interop.

### Motivation

Currently, defining a new FFI-visible type requires either writing C++ (`ObjectDef<T>`) or using `@c_class` which expects C++ reflection metadata. `@py_class` removes the C++ requirement entirely: annotate fields with Python types, and the decorator handles type-index allocation, field registration, and dunder generation.

### Usage

```python
from tvm_ffi.dataclasses import py_class, field
from tvm_ffi import Object

@py_class
class Point(Object):
    x: float
    y: float

@py_class("my.namespace.Line", eq=True)
class Line(Object):
    start: Point
    end: Point
    label: str = "default"
```

## Architecture

Registration is split into two phases to handle forward and mutual references:

- **Phase 1** (`_phase1_register_type`): Allocates a C-level type index and inserts the class into the global type registry. Runs eagerly so that self-referential and mutually-referential annotations can be resolved.
- **Phase 2** (`_phase2_register_fields`): Resolves string annotations via `typing.get_type_hints`, converts them to `TypeSchema`/`Field` objects, validates field ordering, registers fields with the Cython layer, and installs `__init__`/`__repr__`/`__eq__`/etc.

When `get_type_hints` raises `NameError` (forward reference not yet defined), the class is deferred to a pending list and retried after each successful phase-2. A temporary `__init__` triggers lazy completion on first instantiation. Failed phase-2 rolls back phase-1 so the type key can be reused.

## Public Interfaces

| Symbol | Location | Description |
|--------|----------|-------------|
| `@py_class` | `tvm_ffi.dataclasses.py_class` | Main decorator (bare, string-key, and kwarg calling conventions) |
| Parameters | `type_key`, `init`, `repr`, `eq`, `order`, `unsafe_hash`, `kw_only`, `slots` | Mirrors `dataclasses.dataclass` semantics |

## Test Plan

- [x] 246 tests across 47 test classes (`tests/python/test_dataclass_py_class.py`, 3533 lines)
- [x] Coverage includes: basic registration, field parsing, defaults/factories, kw_only, ClassVar, init generation, post_init, repr, equality, ordering, hashing, copy/deepcopy, single/multi-level inheritance, forward/mutual references, dunder preservation, registration rollback, type conversion errors, getter/setter corner cases, memory lifetime, bool alignment, optional fields, Any fields, FFI global function interop
- [x] All 1817 Python tests pass (25 skipped, 3 xfailed) -- no regressions
- [x] All pre-commit hooks pass

### Untested Edge Cases

- Thread safety of the pending-class flush mechanism
- Interaction with `importlib.reload()` on modules containing `@py_class`-decorated types
- Deeply nested generic type annotations (e.g. `List[Optional[Map[str, Array[int]]]]`)

### CODE REVIEW

File: python/tvm_ffi/dataclasses/py_class.py [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+#: forward reference.  Retried after each successful phase-2 via
+#: :func:`_flush_pending`.
+_PENDING_CLASSES: list[_PendingClass] = []
+
+#: Per-module mapping of ``class.__name__ → class`` for every
+#: ``@py_class``-decorated type.  Used as *localns* when resolving
+#: annotations so that mutual references between classes in the same
+#: module work even before the second class is assigned to the module
+#: variable by Python.
+_PY_CLASS_BY_MODULE: dict[str, dict[str, type]] = {}
```
[the remark this answers is not in the document]
Comment (junrushao) [reply python/tvm_ffi/dataclasses/py_class.py:85]: Thanks for the review! This is a valid observation. However, Python's GIL already serializes bytecode execution, and `@py_class` decoration happens at module-import time which is inherently single-threaded per module. Adding `threading.Lock` + double-checked locking would add complexity for a scenario that doesn't arise in practice (dynamically defining `@py_class` classes from multiple threads simultaneously). This is already called out explicitly in the PR description under "Untested Edge Cases." We can revisit if a real use case emerges.

File: python/tvm_ffi/dataclasses/py_class.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+            f = class_val
+        elif class_val is not MISSING:
+            f = field(default=class_val)
+        else:
+            f = field()
+        if class_val is not MISSING:
+            try:
+                delattr(cls, name)
+            except AttributeError:
+                pass
```
[the remark this answers is not in the document]
Comment (junrushao) [reply python/tvm_ffi/dataclasses/py_class.py:186]: Good catch on the redundancy. The `try/except AttributeError` is a defensive pattern against exotic metaclasses or descriptors with custom `__delete__` that might raise unexpectedly. It's only 3 extra lines per site and harmless, so I'd prefer to keep it for robustness. Removing it saves little but introduces a small risk under edge-case metaclass usage.
